src/rate_limiter.rs

```diff
@@ -0,0 +1,224 @@
+// bound by the terms of the MaidSafe Contributor Agreement.  This, along with the Licenses can be
+// found in the root directory of this project at LICENSE, COPYING and CONTRIBUTOR.
+//
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+use error::RoutingError;
+#[cfg(feature = "use-mock-crust")]
+use fake_clock::FakeClock as Instant;
+use maidsafe_utilities::serialisation::{self, SerialisationError};
+use messages::UserMessage;
+use std::cmp;
+use std::collections::BTreeMap;
+use std::net::IpAddr;
+#[cfg(not(feature = "use-mock-crust"))]
+use std::time::Instant;
+
+/// Maximum total bytes the `RateLimiter` allows at any given moment.
+const CAPACITY: u64 = 20 * 1024 * 1024;
+/// The number of bytes per second the `RateLimiter` will "leak".
+const RATE: f64 = 20.0 * 1024.0 * 1024.0;
+/// Charge (in bytes) for a client get request.
+const CLIENT_GET_CHARGE: u64 = 2 * 1024 * 1024;
+
+/// Used to throttle the rate at which clients can send messages via this node. It works on a "leaky
+/// bucket" principle: there is a set rate at which bytes will leak out of the bucket, there is a
+/// maximum capacity for the bucket, and connected clients each get an equal share of this capacity.
+#[derive(Debug)]
+pub struct RateLimiter {
+    /// Map of client IP address to their total bytes remaining in the `RateLimiter`.
+    used: BTreeMap<IpAddr, u64>,
+    /// Timestamp of when the `RateLimiter` was last updated.
+    last_updated: Instant,
+}
+
+impl RateLimiter {
+    pub fn new() -> Self {
+        RateLimiter {
+            used: BTreeMap::new(),
+            last_updated: Instant::now(),
+        }
+    }
+
+    /// Try to add a message. If the message is a form of get request, `CLIENT_GET_CHARGE` bytes
+    /// will be used, otherwise the actual length of the `payload` will be used. If adding that
+    /// amount will cause the client to exceed its share of the `CAPACITY` or cause the total
+    /// `CAPACITY` to be exceeded, `Err(ExceedsRateLimit)` is returned. If the message is invalid,
+    /// `Err(InvalidMessage)` is returned (this probably indicates malicious behaviour).
+    pub fn add_message(&mut self,
+                       online_clients: u64,
+                       client_ip: &IpAddr,
+                       part_count: u32,
+                       part_index: u32,
+                       payload: &[u8])
+                       -> Result<(), RoutingError> {
+        self.update();
+        let total_used: u64 = self.used.values().sum();
+        let used = self.used.get(client_ip).map_or(0, |used| *used);
+        let allowance = cmp::min(CAPACITY - total_used, CAPACITY / online_clients - used);
```

Comment:
Good catch :)  This would be a problem even going from 1 to 2 clients if the first client had taken more than half the capacity!  Again, I'd prefer to leave adding the test case to a separate PR.